### DISCUSSION TITLE
Field Split Preconditioner or Physics Based Preconditioner

### DISCUSSION THREAD

**Discussion by makeclean** [idaholab/moose#20274] [Q&A General] [answered]
I'm doing a comparison against another FE package, and getting some wickedly quick runtimes from it, I think its down to the way it solves one field and then another, rather than solving both. What I can solve in 12 hours, it can solve in 2, for the exact same physics setup. Both use PETSc, so I'm currently assuming its down to the pre-conditioing. It has some interesting PC settings which I've ported across;
```
[Preconditioning]
  [FSP]
    type = FSP
    topsplit = 'temp_disp'
    [temp_disp]
      splitting = 'temp disp'
      splitting_type = additive
    []
    [temp]
      vars = 'temp'
      petsc_options_iname = '-pc_type -pc_hypre_type -ksp_rtol
                             -pc_hypre_boomeramg_strong_threshold -pc_hypre_boomeramg_coarsen_type 
                             -pc_hypre_boomeramg_interp_type -ksp_type
                             -pc_hypre_boomeramg_agg_nl -pc_hypre_boomeramg_agg_num_paths'
      petsc_options_value = ' hypre      boomeramg   1.0e-8            
                              0.75       HMIS                       
                              ext+i cg
                              4  2'

    []
    [disp]
      vars = 'disp_x disp_y disp_z'
      petsc_options_iname = '-pc_type -ksp_type -ksp_rtol 
                             -pc_gamg_type -pc_gamg_agg_nsmooths
                             -pc_gamg_threshold  -pc_gamg_coarse_eq_limit
                             -pc_gamg_square_graph -pc_gamg_reuse_interpolation
                             -mg_levels_esteig_ksp_type -mg_levels_ksp_type
                             -mg_levels_pc_type -mg_levels_ksp_chebyshev_esteig_steps
                             -mg_levels_esteig_ksp_max_it
                             -pc_gamg_sym_graph'
      petsc_options_value = ' gamg     cg         1e-12 
                              agg     1     
                              0.015   1000
                              2       1
                              cg     chebyshev
                              jacobi   20
                              20
                              true'
    []
  []
[]
```
This is my first attempt, but I get the following error
```
[0]PETSC ERROR: --------------------- Error Message --------------------------------------------------------------
[0]PETSC ERROR: No support for this operation for this object type
[0]PETSC ERROR: DM type moose does not implement DMCreateGlobalVector
[0]PETSC ERROR: See https://www.mcs.anl.gov/petsc/documentation/faq.html for trouble shooting.
[0]PETSC ERROR: Petsc Release Version 3.15.1, unknown 
[0]PETSC ERROR: /home/dc-davi4/rds/rds-ukaea-ap001/moose_dev/moose//modules/combined/combined-opt on a  named cpu-p-60 by dc-davi4 Thu Feb 10 13:29:05 2022
[0]PETSC ERROR: Configure options --download-hypre=1 --with-shared-libraries=1 --with-debugging=no --download-fblaslapack=1 --download-metis=1 --download-ptscotch=1 --download-parmetis=1 --download-superlu_dist=1 --download-mumps=1 --download-strumpack=1 --download-scalapack=1 --download-slepc=1 --with-mpi=1 --with-openmp=1 --with-cxx-dialect=C++11 --with-fortran-bindings=0 --with-sowing=0 --with-64-bit-indices --prefix=/home/dc-davi4/rds/rds-ukaea-ap001/moose_dev/petsc --with-make-np=24 --prefix=/home/dc-davi4/rds/rds-ukaea-ap001/moose_dev/petsc
[0]PETSC ERROR: #1 DMCreateGlobalVector() at /rds/project/iris_vol2/rds-ukaea-ap001/moose_dev/moose/petsc/src/dm/interface/dm.c:961
[0]PETSC ERROR: #2 DMGetGlobalVector() at /rds/project/iris_vol2/rds-ukaea-ap001/moose_dev/moose/petsc/src/dm/interface/dmget.c:165
[0]PETSC ERROR: #3 KSPCreateVecs() at /rds/project/iris_vol2/rds-ukaea-ap001/moose_dev/moose/petsc/src/ksp/ksp/interface/iterativ.c:1564
[0]PETSC ERROR: #4 PCSetUp_MG() at /rds/project/iris_vol2/rds-ukaea-ap001/moose_dev/moose/petsc/src/ksp/pc/impls/mg/mg.c:1179
[0]PETSC ERROR: #5 PCSetUp_GAMG() at /rds/project/iris_vol2/rds-ukaea-ap001/moose_dev/moose/petsc/src/ksp/pc/impls/gamg/gamg.c:803
[0]PETSC ERROR: #6 PCSetUp() at /rds/project/iris_vol2/rds-ukaea-ap001/moose_dev/moose/petsc/src/ksp/pc/interface/precon.c:1015
[0]PETSC ERROR: #7 KSPSetUp() at /rds/project/iris_vol2/rds-ukaea-ap001/moose_dev/moose/petsc/src/ksp/ksp/interface/itfunc.c:406
[0]PETSC ERROR: #8 KSPSolve_Private() at /rds/project/iris_vol2/rds-ukaea-ap001/moose_dev/moose/petsc/src/ksp/ksp/interface/itfunc.c:850
[0]PETSC ERROR: #9 KSPSolve() at /rds/project/iris_vol2/rds-ukaea-ap001/moose_dev/moose/petsc/src/ksp/ksp/interface/itfunc.c:1085
[0]PETSC ERROR: #10 PCApply_FieldSplit() at /rds/project/iris_vol2/rds-ukaea-ap001/moose_dev/moose/petsc/src/ksp/pc/impls/fieldsplit/fieldsplit.c:1252
[0]PETSC ERROR: #11 PCApply() at /rds/project/iris_vol2/rds-ukaea-ap001/moose_dev/moose/petsc/src/ksp/pc/interface/precon.c:444
[0]PETSC ERROR: #12 PCApplyBAorAB() at /rds/project/iris_vol2/rds-ukaea-ap001/moose_dev/moose/petsc/src/ksp/pc/interface/precon.c:725
[0]PETSC ERROR: #13 KSP_PCApplyBAorAB() at /home/dc-davi4/rds/rds-ukaea-ap001/moose_dev/moose/petsc/include/petsc/private/kspimpl.h:393
[0]PETSC ERROR: #14 KSPGMRESCycle() at /rds/project/iris_vol2/rds-ukaea-ap001/moose_dev/moose/petsc/src/ksp/ksp/impls/gmres/gmres.c:163
[0]PETSC ERROR: #15 KSPSolve_GMRES() at /rds/project/iris_vol2/rds-ukaea-ap001/moose_dev/moose/petsc/src/ksp/ksp/impls/gmres/gmres.c:249
[0]PETSC ERROR: #16 KSPSolve_Private() at /rds/project/iris_vol2/rds-ukaea-ap001/moose_dev/moose/petsc/src/ksp/ksp/interface/itfunc.c:912
[0]PETSC ERROR: #17 KSPSolve() at /rds/project/iris_vol2/rds-ukaea-ap001/moose_dev/moose/petsc/src/ksp/ksp/interface/itfunc.c:1085
[0]PETSC ERROR: #18 SNESSolve_NEWTONLS() at /rds/project/iris_vol2/rds-ukaea-ap001/moose_dev/moose/petsc/src/snes/impls/ls/ls.c:225
[0]PETSC ERROR: #19 SNESSolve() at /rds/project/iris_vol2/rds-ukaea-ap001/moose_dev/moose/petsc/src/snes/interface/snes.c:4653
------------------------------------------------------
```
I had a quick look at Physics Based Precondition (PBP) but the documentation is a bit scarce, https://mooseframework.inl.gov/source/preconditioners/PhysicsBasedPreconditioner.html. Would I be better off doing a thermal solve, then pass that thermal solution to a subapp to solve the mechanical part of the problem? Or can someone hint at the right way to set this up as a PBP?

**Comment by GiudGiud** [comment]
Hello

I added more docs to the PBP in this PR: https://github.com/idaholab/moose/pull/19985/files
It's not merged yet.

@fdkong have you seen this petsc error?

Guillaume

**Reply by fdkong** [reply]
It is a weird error. GAMG is algebraic and it should not care about DM. 

@makeclean 

Do you mind sharing a simple input file with us on this? I will track it down to reproduce the issue. Regarding the performance, there are many options that can effect performance.   If the other FE code is using PETSc solvers, could you share the results of "-snes_view" if the other FE code uses nonlinear solvers, otherwise "-ksp_view".

With that results, we could setup the same thing for you at the MOSOE side.

**Reply by makeclean** [reply]
The thermal KSP
```
KSP Object: (nls_solve_) 32 MPI processes
  type: cg
  maximum iterations=10000, initial guess is zero
  tolerances:  relative=1e-08, absolute=1e-50, divergence=10000.
  left preconditioning
  using DEFAULT norm type for convergence test
PC Object: (nls_solve_) 32 MPI processes
  type: hypre
  PC has not been set up so information may be incomplete
    HYPRE BoomerAMG preconditioning
      Cycle type V
      Maximum number of levels 25
      Maximum number of iterations PER hypre call 1
      Convergence tolerance PER hypre call 0.
      Threshold for strong coupling 0.75
      Interpolation truncation factor 0.
      Interpolation: max elements per row 0
      Number of levels of aggressive coarsening 4
      Number of paths for aggressive coarsening 2
      Maximum row sums 0.9
      Sweeps down         1
      Sweeps up           1
      Sweeps on coarse    1
      Relax down          symmetric-SOR/Jacobi
      Relax up            symmetric-SOR/Jacobi
      Relax on coarse     Gaussian-elimination
      Relax weight  (all)      1.
      Outer relax weight (all) 1.
      Using CF-relaxation
      Not using more complex smoothers.
      Measure type        local
      Coarsen type        Falgout
      Interpolation type  classical
```
The mechanical
```
KSP Object: (ksp_u_) 32 MPI processes
  type: cg
  maximum iterations=10000, initial guess is zero
  tolerances:  relative=1e-08, absolute=1e-50, divergence=10000.
  left preconditioning
  using DEFAULT norm type for convergence test
PC Object: (ksp_u_) 32 MPI processes
  type: gamg
  PC has not been set up so information may be incomplete
    type is MULTIPLICATIVE, levels=0 cycles=unknown
      Cycles per PCApply=0
      Using externally compute Galerkin coarse grid matrices
      GAMG specific options
        Threshold for dropping small values in graph on each level =  
        Threshold scaling factor for each level not specified = 1.
        AGG specific options
          Symmetric graph false
          Number of levels to square graph 2
          Number smoothing steps 1
        Complexity:    grid = 0.
  linear system matrix = precond matrix:
  Mat Object: 32 MPI processes
    type: mpiaij
    rows=14739, cols=14739, bs=3
    total: nonzeros=602649, allocated nonzeros=602649
    total number of mallocs used during MatSetValues calls=0
      has attached near null space
      using I-node (on process 0) routines: found 158 nodes, limit used is 5
```
I'll try and make a MWE

**Reply by makeclean** [reply]
[sample.zip](https://github.com/idaholab/moose/files/8096341/sample.zip)

Here is the minimum working example

**Reply by fdkong** [reply]
Thanks for example. I will look into it

**Reply by fdkong** [reply]
OK, I looked into this a bit. The issue was that GAMG would use DM to create global vectors when DM is detected since GMAG is built on the top of PCMG. Right now, our DM interface in MOOSE is incomplete. In the current use case, we have to involve DM since FSP is set up via DM.


The short-term solution is to use HYPRE instead of GAMG. I believe you will have a similar performance by replacing GAMG with hypre in FSP.


The long-term solution is to enhance the DM interface in MOOSE https://github.com/idaholab/moose/issues/20420. This will take a while since the team is busy ...

**Reply by makeclean** [reply]
Thanks for taking a look into it @fdkong - is could I use PBP instead would that make a difference? Effectively what I'm trying to head towards is a system where I can solve the thermal system with one preconditioner and the (coupled) structural one with another. I could of course do a thermal solve and have my mechanical solve as a subapp, but is there a neater way to do it all in one input?

**Reply by fdkong** [reply]
>Effectively what I'm trying to head towards is a system where I can solve the thermal system with one preconditioner and the (coupled) structural one with another. 

I meant: you could just replace the GAMG block with HYPRE, and then FSP will work

**Comment by fdkong** [comment, marked as answer]
Do something like this:

```
[Preconditioning]
  [FSP]
    type = FSP
    topsplit = 'temp_disp'
    [temp_disp]
      splitting = 'temp disp'
      splitting_type = additive
    []
    [temp]
      vars = 'temp'
      petsc_options_iname = '-pc_type -pc_hypre_type -ksp_rtol
                             -pc_hypre_boomeramg_strong_threshold -pc_hypre_boomeramg_coarsen_type 
                             -pc_hypre_boomeramg_interp_type -ksp_type
                             -pc_hypre_boomeramg_agg_nl -pc_hypre_boomeramg_agg_num_paths'
      petsc_options_value = ' hypre      boomeramg   1.0e-8            
                              0.75       HMIS                       
                              ext+i cg
                              4  2'

    []
    [disp]
      vars = 'disp_x disp_y disp_z'
        petsc_options_iname = '-pc_type -pc_hypre_type -ksp_rtol
                             -pc_hypre_boomeramg_strong_threshold -pc_hypre_boomeramg_coarsen_type 
                             -pc_hypre_boomeramg_interp_type
                             -pc_hypre_boomeramg_agg_nl -pc_hypre_boomeramg_agg_num_paths'
      petsc_options_value = ' hypre      boomeramg   1.0e-8            
                              0.75       HMIS                       
                              ext+i
                              4  2'
    []
  []
[]
```

**Reply by makeclean** [reply]
Sure, I'll give that a go, but if I wanted to specifically test GAMG then subapp is the way to go? even with a transient problem?

**Reply by fdkong** [reply]
You might have more details about what is your motivation here?  So I understand why you want to do this? If you want to do comparison with other codes,  you might use the same coupling approach as other codes do.

**Reply by makeclean** [reply]
I had a student code up an equivalent physics modules in Fenics/X (https://fenicsproject.org/) and relative to my MOOSE version of the problem, it ran 5x faster. I'm trying to isolate where and why, my initial guess was that it was due to a lot of PC choices I hadn't seen, which was my first go here - to try with the FSP to achieve the same thing. The next nearest thing I can do is a thermal solve, pass that data do the mechanical subapp using the temp as an aux field. I'm trying to achieve similar scaling to his problem, for example, with my full of the problem I can solve 300 timesteps in 12 hours, which I was pretty happy with, but to get that performance I used 360 MPI tasks, 6 tasks per socket which equates to a total of 3360 CPUS reserved (360 being used) across 60 nodes. He was able to do the same with 5 nodes using fully utilised machine, doing all timesteps in 2.5 hours. 

This works out to be 23 seconds per fully converged NL cycle in FENICS and over 375 seconds with MOOSE, using my original PC settings. Fenics doesnt update the mesh, this incurs a cost in time. Maybe I should try the initial precondition only. So thats the motivation here, to figure out what the source of this performance gulf is.

**Reply by fdkong** [reply]
Sorry for late reply again. 

> I'm trying to isolate where and why, my initial guess was that it was due to a lot of PC choices I hadn't seen, which was my first go here - to try with the FSP to achieve the same thing. 

FENICS uses PETSc as well. So the first step would be that:

Run the "FENICS" code with the petsc log option "-log_view".

Run the MOOSE code possibly with FSP (HYPRE for both heat conduction and solid mechanics) with the petsc log option "-log_view" and also moose [pperfgraph](https://mooseframework.inl.gov/source/utils/PerfGraph.html) . 

With these data, we should be able to have a big picture of where differences come from.

The first step of debugging performance is "profiling". Once you have the results of "-log_view", we will have a better idea. 

Put "-log_view" as part of your command line.

**Reply by hugary1995** [reply]
@makeclean  My understanding is that your student is solving the two systems separately, without even using fixed point iteration to get a truly converged solution. Please correct me if I'm wrong.

If my understanding is correct, then in order to make an apples-to-apples comparison, you have to use the MultiApp system in MOOSE to setup the two systems individually (without fixed point iteration). I don't think playing with the FSP will give you a completely decoupled system in the sense that you assemble and solve two matrices separately. Again, @makeclean and @fdkong correct me if I'm wrong.

**Reply by hugary1995** [reply]
I guess another quick way to check is comparing your MOOSE solution to the FENICS solution.

**Reply by fdkong** [reply]
It is a good point. I am not sure how the physics components are coupled at FENICS.  There could be many things there. @makeclean  Do you have PJFNK at FENICS as well? If it is a pure Newton with an exact matrix, we definitely will see huge performance differences. 

It was why I am interested in the results of "-log_view" that will give us a lot of information
